xlcalculator/xlfunctions/financial.py

```diff
@@ -193,6 +220,66 @@ def VDB(
     return result
 
 
+@xl.register()
+@xl.validate_args
+def XIRR(
+        values: func_xltypes.XlArray,
+        dates: func_xltypes.XlArray,
+        guess: func_xltypes.XlNumber
+) -> func_xltypes.XlNumber:
+    """Returns the internal rate of return for a schedule of cash flows that
+        is not necessarily periodic.
+
+    https://support.microsoft.com/en-us/office/
+        xirr-function-de1242ec-6477-445b-b11b-a303ad9adc9d
+
+    Algorithm found on stackoverflow:
+    https://stackoverflow.com/questions/63797804/
+        python-irr-function-giving-different-result-than-excel-xirr
+
+    From MS, Newton's method is used to optimize:
+    https://docs.microsoft.com/en-us/office/troubleshoot/excel/
+        algorithm-of-xirr-funcation
+    """
+    def xnpv(rate, values, dates):
+        if rate <= -1.0:
+            return float('inf')
+
+        min_date = min(dates)
+
+        return sum([value / (1 + rate)**((date - min_date).days / 365)
+                   for value, date in zip(values, dates)])
+
+    def xirr(values, dates, guess):
+        try:
+            return newton(lambda r: xnpv(r, values, dates), guess, maxiter=100)
+
+        except (RuntimeError, FloatingPointError):
+            raise xlerrors.NumExcelError('XIRR did not converge')
+
+    values = values.flatten(func_xltypes.Number, None)
+    dates = dates.flatten(func_xltypes.DateTime, None)
+    values = [float(value) for value in values]
+    dates = [utils.XlDateTime_to_datetime(date) for date in dates]
+    guess = float(guess)
+
+    test = pd.DataFrame({"dates": dates, "values": values})
+
+    # Filter all rows with 0 cashflows
+    test = test[test['values'] != 0]
```

Comment:
I would not call this a test. The values really represent money flow and various points in time.

(At Shoobx we are calculating XIRR for investors, so I have worked with it a lot.)

Reply:
crap.

copy+pasted code and didn't stand back to re-name.

Reply:
I am looking for non-converge scenarios if you've got any laying around...

Reply:
I think we do. I can also give you the exceptions we have gotten and protect against.

BTW, non-convergent is all negative amounts. Also if you have a super large return very close to the investment, the XIRR will be infinite.

I will E-mail you our Excel-compatible XIRR implementation.